providers/implementations/keymgmt/mlkem_kmgmt.c

```diff
@@ -513,135 +487,122 @@ static const OSSL_PARAM *mlkem_gen_settable_params(ossl_unused void *genctx,
 static void *mlkem_gen(void *vctx, OSSL_CALLBACK *osslcb, void *cbarg)
 {
     struct mlkem_gen_ctx *gctx = (struct mlkem_gen_ctx *)vctx;
-    MLKEM768_KEY *mkey;
+    MLKEM_KEY *key;
+    uint8_t *nopub = NULL;
+    uint8_t *seed = gctx->seed;
+    size_t seedlen = MLKEM_SEED_BYTES;
+    vinfo_t v = gctx->vinfo;
+    int genok = 0;
 
-    debug_print("MLKEMKM gen called for %p\n", gctx);
     if (gctx == NULL)
         return NULL;
 
-    if ((mkey = mlkem_new(gctx->provctx)) == NULL) {
+    key = mlkem_new(gctx->provctx, gctx->vinfo);
+    if (key == NULL) {
         ERR_raise(ERR_LIB_PROV, ERR_R_INTERNAL_ERROR);
         return NULL;
     }
 
-    /* If we're doing parameter generation then we just return a blank key */
-    if ((gctx->selection & OSSL_KEYMGMT_SELECT_KEYPAIR) == 0) {
-        debug_print("MLKEMKM gen returns blank %p\n", mkey);
-        return mkey;
-    }
+    /* Actual keypair generation may optionally be deferred */
+    if ((gctx->selection & OSSL_KEYMGMT_SELECT_KEYPAIR) == 0)
+        return key;
 
-    mkey->keytype = MLKEM_KEY_TYPE_768;
+    genok = seed != NULL ?
+        ossl_mlkem_vgenkey_seed(v, seed, seedlen, nopub, 0, &key->prvkey,
```

Comment:
Yes, I want to retain the option of returning the seed, because we may need it if/when someone decides that the private key is `(d, z)`, and then we'll need to retain the generated seed. We may then need another PARAM to specify the preferred private key format (specified prior to import/export).